Advertise NODE_SEGWIT2X service flag + peer seeking.

jgarzik opened the pull request.

jgarzik committed:
Advertise NODE_SEGWIT2X service flag by default.

christophebiocca commented on src/init.cpp:
```diff
@@ -1048,6 +1048,9 @@ bool AppInitParameterInteraction()
     fAcceptDatacarrier = GetBoolArg("-datacarrier", DEFAULT_ACCEPT_DATACARRIER);
     nMaxDatacarrierBytes = GetArg("-datacarriersize", nMaxDatacarrierBytes);
 
+    nLocalServices = ServiceFlags(nLocalServices | NODE_SEGWIT2X);
```
Question: Should this flag instead be set if and only if `NODE_WITNESS` is set (see code on line 1579)?

I don't know the answer as it depends on what the flag is used for.

jgarzik replied:
Matching the conditionality of `NODE_WITNESS` is a slightly more delicate operation imo, wrt tests.

What used for?   Good q.   The immediate goal is visibility to DNS seeds and AddrMan peers via peer exchange.  Once this is seeded out in the field, we can (or might not need to) take further steps to discover this subset of the network.  So, this is a step in a direction, beyond seeds, where there is as yet unscoped second step re peer rendezvous.

christophebiocca replied:
Right, for preferential peering hard-setting the bit is definitely the simplest solution.

tomasvdw commented:
I think that trying to manipulate the network topology through filtered DNS seeding is futile and wrong.

The vast majority of relevant nodes already knows peers and is unaffected by this. The idea that DNS seeding alone will change connectivity ignores both existing peer databases and via-peer address acquisition. 

If certain important nodes need to be connected  (as per the popped up issue), manual override is much more effective.

Seeds must always be diverse and should not be used to attempt to dictate network topology.

jgarzik commented:
Take a look at `NODE_WITNESS` however.  You have a subset of nodes which need to find each other to traffic in "real" blocks - ones with witness data - a self partitioning of the network.

Discovering peers with matching consensus rules is definitely aligned with the 2M HF event down the road.

tomasvdw commented:
Fair point. Although NODE_WITNESS is used for p2p message compatibility and this flag isn't, and although I think the attempt to increase connectivity through seeding is futile, this could certainly help down the bumpy road ahead.

christophebiocca commented:
> I think the attempt to increase connectivity through seeding is futile.

The time that that really matters is at initial setup, post HF. It allows a new node to get a list of nodes that it won't immediately ban, and which won't immediately ban *it*, for sending "bad" blocks.

In addition, once the hard fork sets in there will be 2 fully-partitioned networks, (and banned nodes don't show up in `addr` messages IIRC), so you'll want a peer on the "right" side or you'll be very stuck.

jgarzik committed:
segwit2x node preferential peering policy

1) Advertise node_segwit2x
2) Prefer to peer with segwit2x and, for now, segwit nodes
3) Option to disable pref peering

jgarzik commented:
Updated with preferred peering policy.

jgarzik commented:
@tomasvdw It's not just DNS seeding.  It's everything that `nRelevantServices` does.
